fraud/service.go

```diff
@@ -0,0 +1,192 @@
+package fraud
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/celestiaorg/celestia-node/header"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+)
+
+// headerFetcher aliases a function that is used to fetch ExtendedHeader from store
+type headerFetcher func(context.Context, uint64) (*header.ExtendedHeader, error)
+
+// topics allows to operate with pubsub connection and pubsub topics
+type topics struct {
+	pubsub       *pubsub.PubSub
+	pubSubTopics map[ProofType]*pubsub.Topic
+	mu           sync.RWMutex
+}
+
+// getTopic joins a pubsub.Topic if it was not joined before and returns it
+func (t *topics) getTopic(proofType ProofType) (*pubsub.Topic, bool, error) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	topic, ok := t.pubSubTopics[proofType]
+	if !ok {
+		var err error
+		topic, err = t.pubsub.Join(getSubTopic(proofType))
+		if err != nil {
+			return nil, ok, err
+		}
+		log.Debugf("successfully subscibed to topic", getSubTopic(proofType))
+		t.pubSubTopics[proofType] = topic
+	}
+
+	return topic, ok, nil
+}
+
+// publish allows to publish Fraud Proofs to the network
+func (t *topics) publish(ctx context.Context, data []byte, proofType ProofType) error {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+
+	if topic, ok := t.pubSubTopics[proofType]; ok {
+		return topic.Publish(ctx, data)
+	}
+
+	return errors.New("topic is not found")
+}
+
+// service is propagating and validating Fraud Proofs
+// service implements Service interface.
+type service struct {
+	topics       *topics
+	unmarshalers map[ProofType]ProofUnmarshaler
+	getter       headerFetcher
+	mu           sync.RWMutex
+}
+
+func NewService(p *pubsub.PubSub, getter headerFetcher) Service {
+	return &service{
+		topics: &topics{
+			pubsub:       p,
+			pubSubTopics: make(map[ProofType]*pubsub.Topic),
+		},
+		unmarshalers: make(map[ProofType]ProofUnmarshaler),
+		getter:       getter,
+	}
+}
+
+func (f *service) Start(context.Context) error {
+	log.Info("fraud service is starting...")
+
+	err := f.RegisterUnmarshaler(BadEncoding, UnmarshalBEFP)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (f *service) Stop(context.Context) error {
+	log.Info("fraud service is stopping...")
+	return nil
+}
+
+func (f *service) Subscribe(proofType ProofType) (Subscription, error) {
+	// TODO: @vgonkivs check if fraud proof is in fraud store, then return with error
+	f.mu.RLock()
+	u, ok := f.unmarshalers[proofType]
+	f.mu.RUnlock()
+	if !ok {
+		return nil, errors.New("unmarshaler is not registered")
```

Comment:
```suggestion
		return nil, errors.New("fraud: unmarshaler is not registered")
```
Prepending pkg name simplfies debugging errors

Reply:
package was added in interface.go : `var log = logging.Logger("fraud")`

Reply:
Yes, sorry. Missed that